src/feature/dircache/dircache_stub.c

```diff
@@ -0,0 +1,152 @@
+/* Copyright (c) 2001-2004, Roger Dingledine.
+ * Copyright (c) 2004-2006, Roger Dingledine, Nick Mathewson.
+ * Copyright (c) 2007-2020, The Tor Project, Inc. */
+/* See LICENSE for licensing information */
+
+/**
+ * @file dircache_stub.c
+ * @brief Stub declarations for use when dircache module is disabled.
+ **/
+
+#include "core/or/or.h"
+#include "feature/dircache/consdiffmgr.h"
+#include "feature/dircache/dircache.h"
+#include "feature/dircache/dirserv.h"
+#include "feature/dircommon/dir_connection_st.h"
+
+int
+directory_handle_command(dir_connection_t *conn)
+{
+  (void) conn;
+  tor_assert_nonfatal_unreached_once();
+  return -1;
+}
+
+int
+connection_dirserv_flushed_some(dir_connection_t *conn)
+{
+  (void) conn;
+  tor_assert_nonfatal_unreached_once();
+  return -1;
+}
+
+int
+directory_fetches_from_authorities(const or_options_t *options)
+{
+  (void) options;
+  return 0;
+}
+
+int
+directory_fetches_dir_info_early(const or_options_t *options)
```

Comment:
`FetchDirInfoEarly` is listed as a general option in the man page. It does work on clients, and it's used by some custom client setups (like bandwidth scanners, and other directory monitors).

I think it's ok to disable this option when dircache mode is disabled, but we need to document that in the man page. Maybe it should be moved to an "Advanced Options" section, with a note that these options don't work on clients, when compiled with dircache mode disabled?

Reply:
(This function calls `directory_fetches_from_authorities()`, which checks the value of `FetchDirInfoEarly`.)

Reply:
These comments also apply to `FetchDirInfoExtraEarly`.

Reply:
There are a bunch of related options used by `should_use_directory_guards()`:
* UseEntryGuards
* DownloadExtraInfo
* FetchDirInfoEarly
* FetchDirInfoExtraEarly
* FetchUselessDescriptors
https://github.com/torproject/tor/blob/4f02812242d1fd90d859eb98ac3fb1ed182f18cf/src/feature/dirclient/dirclient.c#L310

For consistency, we should probably either keep all these options, or disable all of them. If we decide to disable them, let's do it in a separate ticket, and make `directory_fetches_dir_info_early()` a dircommon function for now?